### PR TITLE
Service refactor

### PR DESCRIPTION
**Pull request by mrsmkl**
## Description

Add a description of your changes here.

## Is this PR related with an open issue?

Related to Issue #

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
 
## Checklist:

- [ ] Follows the code style of this project.
- [ ] Tests Cover Changes
- [ ] Documentation

### CODE REVIEW

File: src/access/access.controller.ts [hunk 1]
```diff
@@ -161,107 +121,18 @@ export class AccessController {
...
-  ): Promise<string> {
-    const nevermined = await Nevermined.getInstance(config);
-    const instanceConfig = {
-      ...generateIntantiableConfigFromConfig(config),
-      nevermined
+  async doNftTransfer(@Body() transferData: TransferDto, @Req() req: Request<unknown>): Promise<string> {
+    const nevermined = await getNevermined();
+    const params: ValidationParams = {
+      consumer_address: transferData.nftReceiver,
+      did: (await nevermined.keeper.agreementStoreManager.getAgreement(transferData.agreementId)).did,
```
Comment (r-marques) [inline src/access/access.controller.ts:128]: this call should check if the agreementId exists and if not return the appropriate error message

File: src/auth/auth.service.ts [hunk 2]
```diff
@@ -300,19 +72,22 @@ export class AuthService {
     try {
       payload = jwtEthVerify(clientAssertion);
       // const address = payload.iss;
+      // console.log('got payload', payload);
```
Comment (r-marques) [inline src/auth/auth.service.ts:75]: cleanup

File: src/auth/auth.service.ts [hunk 3]
```diff
@@ -300,19 +72,22 @@ export class AuthService {
...
+
+      const params: ValidationParams = {
+        consumer_address: payload.iss,
+        did: '0x'+(payload.did as string).split(':')[2],
```
Comment (r-marques) [inline src/auth/auth.service.ts:79]: this should use the sdk `didZeroX`

File: src/common/helpers/agreement.ts [hunk 4]
```diff
@@ -18,80 +16,17 @@ async function fetch(...args) {
...
-export interface Params<T> {
-  nevermined: Nevermined,
-  agreement_id: string, 
-  did: string, 
-  template: Template<T>,
-  params: T,
-  conditions: ConditionInfo[]
-}
-
 export async function getNevermined() {
```
Comment (r-marques) [inline src/common/helpers/agreement.ts:19]: nest.js must have a better way to do this with global state. Instantiate this once and share across requests

File: src/common/helpers/agreement.ts [hunk 5]
```diff
@@ -127,10 +63,14 @@ function parseUrl(url: string): string {
...
-export async function downloadAsset(did: string, index: number, res: any): Promise<StreamableFile> {
+export async function downloadAsset(did: string, index: number, res: any): Promise<StreamableFile|string> {
   try {
-    let {url, content_type} = await getAssetUrl(did, index)
+    let {url, content_type, dtp} = await getAssetUrl(did, index)
+    if (dtp) {
+      return url
+    }
     if (!url) {
+      // TODO: add log
```
Comment (r-marques) [inline src/common/helpers/agreement.ts:73]: This is not an internal server exception. We know what the problem is and should return a 4xx with a proper error message

**Comment by r-marques** [comment]
there is both a `package-lock.json` and `yarn.lock` in this repo

File: src/auth/auth.service.ts [hunk 6]
```diff
@@ -24,261 +22,34 @@ export class AuthService {
...
     };
     const dtp = await Dtp.getInstance(instanceConfig);
-    const consumer = await dtp.babyjubPublicAccount('0x'+buyer.substring(0,64), '0x'+buyer.substring(64,128));
-    const { url } = await getAssetUrl(did, 0);
-    const data = Buffer.from(url, 'hex');
-    const extra : AccessProofConditionExtra = {
-      providerK: dtp.keytransfer.makeKey(process.env.PROVIDER_BABYJUB_SECRET),
-      data
-    };
     const buyerPub = new BabyjubPublicKey('0x'+buyer.substring(0,64), '0x'+buyer.substring(64,128));
```
Comment (aaitor) [inline src/auth/auth.service.ts:49]: this should use the sdk `didZeroX`

File: src/access/access.controller.ts (hunk 1 above)
Comment (aaitor) [reply src/access/access.controller.ts]: @mrsmkl , I think this is not addressed yet

File: src/common/helpers/agreement.ts [hunk 7]
```diff
@@ -127,11 +49,15 @@ function parseUrl(url: string): string {
...
+export async function downloadAsset(nvm: Nevermined, did: string, index: number, res: any): Promise<StreamableFile|string> {
   try {
-    let {url, content_type} = await getAssetUrl(did, index)
+    let {url, content_type, dtp} = await getAssetUrl(nvm, did, index)
+    if (dtp) {
+      return url
+    }
     if (!url) {
-      throw new InternalServerErrorException(undefined, 'Bad URL')
+      // TODO: add log
```
[the remark this answers is not in the document]
Comment (mrsmkl) [reply src/common/helpers/agreement.ts:59]: I'd prefer to keep these comments, adding better logging should be the next thing to do.

**Review by aaitor — APPROVED** [review]

**Pull request merged**